### PR TITLE
auto-batching functionality

### PR DESCRIPTION
**Pull request by ShinyMeta**
Added an "pool" to client to hold static auto-batching endpoints.  
Endpoint now can enableAutoBatch() which creates a set for queueing ids for the next batch.

Example:
client.autoBatch('items', 500).get(100)
client.autoBatch('items', 500).get(100)
client.autoBatch('items', 500).get(101)
client.autoBatch('items', 500).many([100,101,102])
client.autoBatch('items', 500).get(101)
client.autoBatch('items', 500).get(101)
//when all of these requests are made in sequence/from multiple locations in code, results in a single api request for 100,101,102, and resolves each one with the requested ids

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -79,6 +80,27 @@ module.exports = class Client {
     })
   }
 
+  //maintains pool of static endpoints with auto-batching enabled
```
Comment (queicherius) [inline src/client.js]: Nit: All comments should follow the schema of a grammatically correct sentence:

```suggestion
// Maintains a pool of static endpoints with auto-batching enabled
```

(Here and others)

File: src/endpoint.js [hunk 2]
```diff
@@ -19,12 +19,14 @@ module.exports = class AbstractEndpoint {
     this.isPaginated = false
     this.maxPageSize = 200
     this.isBulk = false
-    this.supportsBulkAll = true
+    this.supportsBulkAll = true
```
Comment (queicherius) [inline src/endpoint.js]: Remove the whitespace (the linter will complain about this too if you run `npm run lint` / `yarn lint`)

File: src/endpoint.js [hunk 3]
```diff
@@ -74,6 +76,19 @@ module.exports = class AbstractEndpoint {
     return this
   }
 
+  //turns on auto-batching for this endpoint
+  enableAutoBatch (interval) {
+    if (this._autoBatch === null) {
+      this._autoBatch = {
+        interval: interval || 1000,
```
Comment (queicherius) [inline src/endpoint.js]: - Maybe rename into `windowSize`? Interval means more that it runs in that specific interval even it if didnt get called.
- This feels like a pretty high default to me, since it adds 1s to each batched request (that takes <200ms usually). 50ms or something like that seems more reasonable.

File: src/endpoint.js [hunk 4]
```diff
@@ -74,6 +76,19 @@ module.exports = class AbstractEndpoint {
...
+        set: new Set(),
```
Comment (queicherius) [inline src/endpoint.js]: Rename into something that explains better what it does, e.g. `ids`

File: src/endpoint.js [hunk 5]
```diff
@@ -74,6 +76,19 @@ module.exports = class AbstractEndpoint {
...
+        nextBatchPromise: null,
+        autoBatchOverride: false,
```
Comment (queicherius) [inline src/endpoint.js]: Remove this - it's essentially global state, has problems with race conditions (something might call `many` right before the batch request hits) and it makes it pretty hard to follow what's going on. Instead add an additional parameter like `skipAutoBatch` to `many`/`_many` and pass `true` through when the batch operation calls `many`.

File: src/endpoint.js [hunk 6]
```diff
@@ -168,6 +190,36 @@ module.exports = class AbstractEndpoint {
     return this._request(this.url)
   }
 
+  _autoBatchMany (ids) {
+    if (!this._autoBatch.set) {
+      this._autoBatch.set = new Set()
+    }
```
Comment (queicherius) [inline src/endpoint.js]: This can never happen

File: src/endpoint.js [hunk 7]
```diff
@@ -168,6 +190,36 @@ module.exports = class AbstractEndpoint {
...
+    if (this._autoBatch.set.size === 0) {
+      this._autoBatch.nextBatchPromise = new Promise((resolve, reject) => {
+        setTimeout(() => {
+          const batchedIds = Array.from(this._autoBatch.set)
+          this.debugMessage(`batch sending for ${batchedIds}`)
```
Comment (queicherius) [inline src/endpoint.js]:
```suggestion
          this.debugMessage(`autoBatchMany called (${batchedIds.length} ids)`)
```

File: src/endpoint.js [hunk 8]
```diff
@@ -168,6 +190,36 @@ module.exports = class AbstractEndpoint {
...
+          this._autoBatch.set.clear()
+          this._autoBatch.autoBatchOverride = true
+          return resolve(this.many(batchedIds))
+        }, this._autoBatch.interval) // by default is 1000, 1 sec
```
Comment (queicherius) [inline src/endpoint.js]: Remove the comment here, because it does not add any value

File: src/endpoint.js [hunk 9]
```diff
@@ -168,6 +190,36 @@ module.exports = class AbstractEndpoint {
...
+      }).then(items => {
+        const indexedItems = {}
+        items.forEach(item => {
+          indexedItems[item.id] = item
+        })
+        return indexedItems
+      })
+    }
+
+    //add ids to set
```
Comment (queicherius) [inline src/endpoint.js]:
```suggestion
    // Add the requested ids to the pending ids
```

File: src/endpoint.js [hunk 10]
```diff
@@ -168,6 +190,36 @@ module.exports = class AbstractEndpoint {
...
+    ids.forEach(id => this._autoBatch.set.add(id))
+    // return array with results requested
```
Comment (queicherius) [inline src/endpoint.js]:
```suggestion
    // Return the results based on the requested ids
```

**Review by queicherius — CHANGES_REQUESTED** [review]
Thank you for your contribution 🙂 I have flagged some issues in the code, please have a look through the comments.

File: src/endpoint.js (hunk 3 above)
Comment (ShinyMeta) [reply src/endpoint.js]: Updated the name to batchDelay, and had updated the delay to 100 before I caught this comment, can reduce to 50 if that makes more sense to you

File: src/endpoint.js (hunk 4 above)
Comment (ShinyMeta) [reply src/endpoint.js]: New name: idsForNextBatch

File: src/endpoint.js (hunk 5 above)
Comment (ShinyMeta) [reply src/endpoint.js]: yep, makes perfect sense, implemented these recommendations

File: src/client.js [hunk 11]
```diff
@@ -79,6 +80,27 @@ module.exports = class Client {
     })
   }
 
+  // Maintains a pool of static endpoints with auto-batching enabled
+  autoBatch (endpointName, batchDelay) {
```
Comment (ShinyMeta) [inline src/client.js]: @queicherius When I was designing this, I wasn't sure of a cleaner way to "toggle on" autobatching with a similar syntax to the rest of the library, something like `client.autobatch().items()` .  Do you have some thoughts on a good way to do that?  I'd like to keep the style of the library consistent.

Comment (ShinyMeta) [reply src/client.js]: as a second note, my implementation doesn't support child endpoints, like account().achievements().  The primary use case I'm imagining is items(), but I'd like it to just be a feature everywhere that it could be useful.

Comment (ShinyMeta) [reply src/client.js]: Actually I have an idea for how to do this with an autobatchPool class returned from client.autoBatch() that would function as a wrapper for the endpoints... This should be a lot cleaner

File: src/endpoint.js (hunk 3 above)
Comment (queicherius) [reply src/endpoint.js]: I feel like 50 is probably the better default, just because 100 is already 50% of just shooting out the request instantly

File: src/client.js (hunk 11 above)
Comment (queicherius) [reply src/client.js]: Right now, it just gets passed down for other things that do chaining (e.g. here's how language works):
- Set language on the client: https://github.com/queicherius/gw2api-client/blob/master/src/client.js#L25-L29
- Inherit the language from the parent (this can be a client or a other endpoint): https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L12
- Set the language on the endpoint: https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L44-L48

**Comment by ShinyMeta** [comment]
I moved most of the logic in Client to a new class, AutoBatchNode, which is essentially a wrapper class for Client/Endpoint.  This changes the syntax to match the current library as follows:

Example:
```
api.autoBatch().items().get(100)
api.autoBatch().items().get(100)
api.autoBatch().items().get(101)
api.autoBatch().items().many([100,101,102])
api.autoBatch().items().get(101)
api.autoBatch().items().get(101)
```

This also works in chaining with settings/child endpoints now:
`api.autoBatch().authenticate('api key').account().achievements().many([1,2,3])`

Also, if it is desired for ALL bulk expanding requests to autobatch, the Autobatch node can be stored and used almost identically to a normal client:
`api = client().autoBatch()`

CONSIDERATIONS:
Since the autobatching endpoints are static, their settings such as apikey and schema should not be changed individually.  Thus there is a line of code to force their settings to match the source client on every subsequent method call:
```
this.parent
      .schema(this.client.schemaVersion)
      .language(this.client.lang)
      .authenticate(this.client.apiKey)
      .debugging(this.client.debug)
```

Auto batching endpoints also override the `_skipCache` setting which is normally changed by `.live()`, to force cache usage.

Default batch delay is 50, but this CAN be changed per endpoint through subsequent calls to the `enableAutoBatch(batchDelay)` method.

The chain after `.autoBatch()` will continue to return AutoBatchNodes wrapping the client/endpoint until a method is called that would return anything other than a Client/Endpoint (i.e. Promise/bool/string).

File: src/endpoint.js [hunk 12]
```diff
@@ -74,6 +77,23 @@ module.exports = class AbstractEndpoint {
     return this
   }
 
+  // Turn on auto-batching for this endpoint
+  enableAutoBatch (autoBatchDelay) {
```
Comment (queicherius) [inline src/endpoint.js:81]: It would be nice to also call this `autoBatch` like in the client, so it's consistent. Similar to `language`, `debugging` etc.

(Expanded on more below)

File: src/endpoint.js [hunk 13]
```diff
@@ -74,6 +77,23 @@ module.exports = class AbstractEndpoint {
...
+    if (this._autoBatch === null) {
+      if (!this.isBulk) {
+        this.debugMessage(`${this.url} is not bulk expanding, endpoint will not have any autobatch behavior`)
+        return this
+      }
+      this._autoBatch = {
+        // autoBatchDelay: autoBatchDelay,
```
Comment (queicherius) [inline src/endpoint.js:88]: Please remove commented out code

File: src/endpoint.js [hunk 14]
```diff
@@ -230,9 +283,17 @@ module.exports = class AbstractEndpoint {
...
   // Get multiple entries by ids from the live API
-  _many (ids, partialRequest = false) {
+  _many (ids, partialRequest = false, skipAutoBatch) {
     this.debugMessage(`many(${this.url}) requesting from api (${ids.length} ids)`)
 
+    if (this._autoBatch !== null) {
+      if (!skipAutoBatch) {
+        return this._autoBatchMany(ids)
+      }
+    }
```
Comment (queicherius) [inline src/endpoint.js:293]:
```suggestion
    if (this._autoBatch !== null && !skipAutoBatch) {
      return this._autoBatchMany(ids)
    }
```

File: src/endpoint.js [hunk 15]
```diff
@@ -230,9 +283,17 @@ module.exports = class AbstractEndpoint {
   }
 
   // Get multiple entries by ids from the live API
-  _many (ids, partialRequest = false) {
+  _many (ids, partialRequest = false, skipAutoBatch) {
```
Comment (queicherius) [inline src/endpoint.js:286]:
```suggestion
  _many (ids, partialRequest = false, skipAutoBatch = false) {
```

File: src/endpoint.js [hunk 16]
```diff
@@ -168,8 +195,34 @@ module.exports = class AbstractEndpoint {
...
+    ids.forEach(id => this._autoBatch.idsForNextBatch.add(id))
+    // Return the results based on the requested ids
+    return this._autoBatch.nextBatchPromise.then(indexedItems => {
+      return ids.map(id => indexedItems[id]).filter(x => x)
+    })
+  }
+
   // Get multiple entries by ids
-  many (ids) {
+  many (ids, skipAutoBatch) {
```
Comment (queicherius) [inline src/endpoint.js:225]:
```suggestion
  many (ids, skipAutoBatch = false) {
```

File: src/autoBatchNode.js [hunk 17]
```diff
@@ -0,0 +1,86 @@
+
+const Endpoint = require('./endpoint')
+const _ = require('lodash')
```
Comment (queicherius) [inline src/autoBatchNode.js:3]: (You don't use it here so it can be removed, but just FYI in 2021 using Lodash is pretty much always an antipattern)

File: src/autoBatchNode.js [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    return childNode
+  }
+
+  _getChildNode(endpointName, args) {
+    const nameList = [endpointName, ...args]
+    const childName = nameList.join('_')
+    return this.children[childName]   
+  }
+
+}
```
Comment (queicherius) [inline src/autoBatchNode.js:86]: I have to say the entire wrapping logic here feels a little over-engineered. What benefits does it have over just passing a `autoBatching = true` flag, just like we pass flags for e.g. `debug`?

- Set the debug flag on the client: https://github.com/queicherius/gw2api-client/blob/master/src/client.js#L45-L49
- Inherit the debug flag from the parent (this can be a client or a other endpoint): https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L16
- Set the debug flag on the endpoint: https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L57-L61
- Use the debug flag: https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L65

It feels like that would ultimately achieve the same functionality and would e.g. also make it possible to pass the `autoBatchDelay` to the client if we did something like this:

```js
// Client

autoBatch (flag = true, autoBatchDelay = 50) {
  this.autoBatching = flag
  this.autoBatchDelay = autoBatchDelay
  return this
}
```

```js
// Endpoint

constructor (parent) {
  // ...
  this.autoBatching = parent.autoBatching
  this.autoBatchDelay = parent.autoBatchDelay

  this._autoBatch = {
    idsForNextBatch: new Set(),
    nextBatchPromise: null,
  }
}

autoBatch (flag = true, autoBatchDelay = 50) {
  // We could also print the debug warning message here, if this.isBulk = false
  // that the flag will just be ignored. But in some cases it might make sense anyway,
  // like this one where it gets passed to a child:
  // client().account().autoBatch(true).characters().many([...])
  // So IMO it would be fine to just set the flag and silently ignore it (see below)
  // if the endpoint is not bulk.

  this.autoBatching = flag
  this.autoBatchDelay = autoBatchDelay
  return this
}

_many (ids, partialRequest = false, skipAutoBatch = false) {
  if (this.isBulk && this.autoBatching && !skipAutoBatch) {
    return this._autoBatchMany(ids)
  }
}
```

```js
// Usage
const api = client().autoBatch(true, 50)
api.items().many([12]) // -> autobatches
api.account().bank().get() // -> doesn't
api.items().autoBatch(false).many([12]) // -> doesn't
client().items().autoBatch(true).get(12) // -> does
```

---

> Since the autobatching endpoints are static, their settings such as apikey and schema should not be changed individually.

I don't think this is an issue, we need to handle. If the user mutates the language on an already existing endpoint with autobatching enabled, while a request is in-flight, I feel like that's on them.

> Auto batching endpoints also override the _skipCache setting which is normally changed by .live(), to force cache usage.

`live()` should win over everything, it is forced to overwrite the cache usage and immediately fire the request.

**Review by queicherius — CHANGES_REQUESTED** [review]

File: src/autoBatchNode.js (hunk 18 above)
Comment (ShinyMeta) [reply src/autoBatchNode.js:86]: I agree it would be cleaner if autobatch were handled via setting flag, however I think it could be confusing. For example:
```javascript
let api = client().autoBatch()
api.items().get(101)
api.items().get(102)
```
This normal behavior of the library is to return a new endpoint from items(), which would have seperate autoBatch sets of ids with my current endpoint proposed edit.  
In my mind, if the client is set to autoBatch, then subsequent calls for the same endpoint should also autoBatch. The main thing the wrapper is doing is caching child endpoints so that subsequent calls are returning the same endpoint objects. For the most part, other method calls are passed through without interference.

The reason I added logic to re-inherit settings from the parent, was probably over-cautious, and unnecessary.  I was worried that people might try to change settings on the endpoint like this:
```javascript
await api.items().live().get(101)
await api.items().get(101)
```
Which is fine with the current library because it makes a new endpoint on each line so the second line will use the cache. But if the api is returning saved endpoints from a pool for autobatching, then the setting stays.  This can be avoided with proper usage of the library, and removing it cuts some fluff from my edit.

Comment (queicherius) [reply src/autoBatchNode.js:86]: > This normal behavior of the library is to return a new endpoint from items(), which would have seperate autoBatch sets of ids with my current endpoint proposed edit. In my mind, if the client is set to autoBatch, then subsequent calls for the same endpoint should also autoBatch.

Ah! I didn't think about batching being shared between endpoints of the same type. That's very valid, and I agree that it should work how you described.

> The main thing the wrapper is doing is caching child endpoints so that subsequent calls are returning the same endpoint objects. For the most part, other method calls are passed through without interference.

Instead of wrapping the client and returning the same endpoint objects, what do you think about a singleton shared data object that holds the ID list and timeout based on the endpoint URL? That should solve the issues around the shared settings and still have the simple interface I suggested.

Something like this?

```js
const autoBatchSharedData = {}

module.exports = class AbstractEndpoint {
  constructor(parent) {
    // ...

    this.setupAutoBatchSharedData()
  }

  // Setup the shared data for auto batching
  setupAutoBatchSharedData() {
    const autoBatchId = this.baseUrl + this.url + ':' + this.schemaVersion
    if (!autoBatchSharedData[autoBatchId]) {
      autoBatchSharedData[autoBatchId] = {
        idsForNextBatch: new Set(),
        nextBatchPromise: null,
      }
    }

    this._autoBatch = autoBatchSharedData[autoBatchId]
  }

  // Set the schema version
  schema(schema) {
    this.schemaVersion = schema
    this.setupAutoBatchSharedData()
    this.debugMessage(`set the schema to ${schema}`)
    return this
  }
}
```

> The reason I added logic to re-inherit settings from the parent, was probably over-cautious, and unnecessary. I was worried that people might try to change settings on the endpoint like this:

```js
await api.items().live().get(101)
await api.items().get(101)
```

Yeah, I would also expect the first one to go through and the second one to hit the cache.

Comment (ShinyMeta) [reply src/autoBatchNode.js:86]: I've been staring at this and wondering why I didn't think of that design before. 

Should language/apikey be added to the batchId with schema or remove schema altogether?
If we care, it should probably match _cacheHash. However this will only cause an issue if the user changes the setting inside of the batch window, which is 50ms by default... Probably just a result of irresponsible use, and could be a consideration in documentation about autoBatching

Comment (queicherius) [reply src/autoBatchNode.js:86]: > Should language/apikey be added to the batchId with schema or remove schema altogether?
If we care, it should probably match _cacheHash. However this will only cause an issue if the user changes the setting inside of the batch window, which is 50ms by default...

I think using `_cacheHash` is a good idea! Then it'd behave correctly in all instances (without race conditions) and the only overhead would be to add the extra function call to the setters for language & api key. Then if a user changes the settings in the batch window it doesnt get batched but it gets the correct result.
